[core] Support pip_install_options for pip

VassilisVassiliadis opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

<!-- Please give a short summary of the change and the problem this solves. -->

The PR introduces support for the field `pip.pip_install_options` for pip which behaves similar to the `uv.uv_pip_install_options` field for the `uv` environment manager.

This will enable us to specify options to the `pip install` command that creates up the virtual environment for my ray job/actors. 

## Related issue number

<!-- For example: "Closes #1234" -->

Closes: #52679

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

VassilisVassiliadis committed:
[core] Support pip_install_options for pip

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

dstrodtman reviewed (approved):
One small suggestion. Did not review runtime_env files as I don't believe these are published as docs, but please LMK if yes.

VassilisVassiliadis committed:
Fix typos in pip_install_options documentation

Co-authored-by: Douglas Strodtman <douglas@anyscale.com>
Signed-off-by: Vassilis Vassiliadis <43679502+VassilisVassiliadis@users.noreply.github.com>

VassilisVassiliadis commented on doc/source/ray-core/handling-dependencies.rst:
```diff
@@ -546,6 +546,7 @@ The ``runtime_env`` is a Python dictionary or a Python class :class:`ray.runtime
   `“requirements.txt” <https://pip.pypa.io/en/stable/user_guide/#requirements-files>`_ file, or (3) a python dictionary that has three fields: (a) ``packages`` (required, List[str]): a list of pip packages,
   (b) ``pip_check`` (optional, bool): whether to enable `pip check <https://pip.pypa.io/en/stable/cli/pip_check/>`_ at the end of pip install, defaults to ``False``.
   (c) ``pip_version`` (optional, str): the version of pip; Ray will spell the package name "pip" in front of the ``pip_version`` to form the final requirement string.
+  (d) ``pip_install_options (optional, List[str]): user-provided options for ``pip install`` command, defaults to ``["--disable-pip-version-check", "--no-cache-dir"]``.
```
Thank you for spotting this @dstrodtman, I completely missed it earlier!

VassilisVassiliadis commented:
Hello, what would be the next step towards merging this PR ?

VassilisVassiliadis committed:
[core][runtime env] unit-test validating pip_install_options with pip

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

VassilisVassiliadis committed:
[core][runtime env] integration-test for using pip_install_options with pip

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

VassilisVassiliadis commented:
Thank you @edoakes, I pushed a couple of commits just now adding one integration test and one unit-test. I tested them on my MacOs machine and they worked as expected.

VassilisVassiliadis committed:
chore run linter on new pip_install_options integration and unit tests

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

edoakes reviewed (approved):
Premerge CI is running now: https://buildkite.com/ray-project/premerge/builds/43265

If it passes, LGTM 🚢

VassilisVassiliadis commented on python/ray/tests/test_runtime_env_conda_and_pip.py:
```diff
@@ -373,5 +373,29 @@ def test_import():
 
+def test_pip_install_options(shutdown_only):
+
+    ray.init()
```
Thank you, I'll remove that line. I basically started from this existing test: https://github.com/ray-project/ray/blob/7ac4083279b8b25ff02ef60df9caa65f93ae69c7/python/ray/tests/test_runtime_env_conda_and_pip.py#L206-L214

would you like me to remove it from  there too ?

edoakes replied:
sure, go for it.

`ray.init()` will be implicitly called on the first `.remote()` call

VassilisVassiliadis replied:
I see, thank you

VassilisVassiliadis committed:
remove unnecessary calls to ray.init() in integration tests for pip

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

edoakes commented:
The added tests are failing: https://buildkite.com/ray-project/premerge/builds/43267#0197ea6f-f8ef-49bf-8fd3-2d867280d9d6

VassilisVassiliadis committed:
[core][runtime env] set default pip_install_options for all variants of a pip runtime environment

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

VassilisVassiliadis committed:
[core][runtime env] update the test_runtime_env_interface test to account for pip_install_options

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

VassilisVassiliadis committed:
[core][runtime env] update the test_job_driver_inheritance test to account for pip_install_options

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

VassilisVassiliadis committed:
[core][runtime env] update the test_job_driver_inheritance_override test to account for pip_install_options

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

VassilisVassiliadis committed:
[core][runtime env] skip defaults for pip.pip_install_options

pip_install_options is a new field. By skipping the insertion of default values during the validation of the pip environment we maintain backwards compatibility with ray==2.0.1

Signed-off-by: Vassilis Vassiliadis <vassilis.vassiliadis@ibm.com>

edoakes reviewed (approved):
Thanks!

VassilisVassiliadis commented:
Awesome, thanks for all the help!
